src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionService.cs

```diff
@@ -0,0 +1,370 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers.ImportCompletion
+{
+    internal abstract partial class AbstractTypeImportCompletionService : ITypeImportCompletionService
+    {
+        private ITypeImportCompletionCacheService CacheService { get; }
+
+        protected abstract string GenericTypeSuffix { get; }
+
+        protected abstract bool IsCaseSensitive { get; }
+
+        internal AbstractTypeImportCompletionService(Workspace workspace)
+        {
+            CacheService = workspace.Services.GetService<ITypeImportCompletionCacheService>();
+        }
+
+        public async Task GetTopLevelTypesAsync(
+            Project project,
+            SyntaxContext syntaxContext,
+            Action<TypeImportCompletionItemInfo> handleItem,
+            CancellationToken cancellationToken)
+        {
+            if (!project.SupportsCompilation)
+            {
+                throw new ArgumentException(nameof(project));
+            }
+
+            var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+            // Since we only need top level types from source, therefore we only care if source symbol checksum changes.
+            var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(project, cancellationToken).ConfigureAwait(false);
+
+            GetAccessibleTopLevelTypesWorker(
+                project.Id,
+                compilation.Assembly,
+                checksum,
+                syntaxContext,
+                handleItem,
+                CacheService.ProjectItemsCache,
+                cancellationToken);
+        }
+
+        public void GetTopLevelTypesFromPEReference(
+            Solution solution,
+            Compilation compilation,
+            PortableExecutableReference peReference,
+            SyntaxContext syntaxContext,
+            Action<TypeImportCompletionItemInfo> handleItem,
+            CancellationToken cancellationToken)
+        {
+            var key = GetReferenceKey(peReference);
+            if (key == null)
+            {
+                // Can't cache items for reference with null key. We don't want risk potential perf regression by 
+                // making those items repeatedly, so simply not returning anything from this assembly, until 
+                // we have a better understanding on this sceanrio.
+                // TODO: Add telemetry
```

Comment:
>// TODO: Add telemetry [](start = 16, length = 22)

Add a github issue

Reply:
I didn't add telemetry here because I was not sure what data can help us understand this case. I can check if we ever  run into this, but w/o knowing the root cause, it's not actionable. Suggestions?